include/sound/soc-acpi.h

```diff
@@ -75,18 +75,47 @@ struct snd_soc_acpi_mach_params {
 };
 
 /**
- * snd_soc_acpi_link_adr: ACPI-based list of _ADR, with a variable
- * number of devices per link
+ * snd_soc_acpi_endpoint: endpoint descriptor
+ *
+ * @num: endpoint number (mandatory, unique per device)
+ * @aggregated: 0 (independent) or 1 (logically grouped)
+ * @group_position: zero-based order (only when @aggregated is 1)
+ * @group_id: platform-unique group identifier (only when @aggregrated is 1)
+ */
+struct snd_soc_acpi_endpoint {
+	u8 num;
+	u8 aggregated;
+	u8 group_position;
+	u8 group_id;
+};
+
+/**
+ * snd_soc_acpi_adr_device: descriptor for device, include _ADR
+ * reresented at 64 bits and endpoint information
+ *
+ * @adr: 64 bit ACPI _ADR value
+ * @num_endpoints: number of endpoints for this device
+ * @endpoints: array of endpoints
+ */
+struct snd_soc_acpi_adr_device {
+	const u64 adr;
+	const u8 num_endpoints;
```

Comment:
@plbossart no objection to the code but for my educatio: num_endpoints doesnt seem to be more than 1 in any of the definitions below. Will it ever be?

Reply:
in most cases there will be a single endpoint per device, at least for the cases where we need to care about aggregation. that said it's probably better to plan for more in case someone combines two audio functions in the same device.
Note that the definition of endpoint is really minimal, e.g. for a headset codec I would most likely need two endpoints. I chose to keep things simple for now, this will be completely revisited anyways for new devices where the machine driver will have to be handled by parsing an ACPI description (similar to DT). The code in this PR is largely a placeholder until we have standardized info in the DSDT.

Reply:
Thanks @plbossart. I suppose I got carried away by the terminology and the definitions a bit. What confused me was the array of snd_soc_acpi_adr_device for each struct snd_soc_acpi_link_adr. It was an array of "u64 adr" even before. But I wonder why do we need an array here? Will there be more than one address per link?

Reply:
yes of course, on TGL RVP we have two RT1308 on the same link. In this case the concept of 'aggregation' is provided directly by SoundWire (same as TDM), so it's way simpler than when two links are used.

Reply:
ahh got it! Sorry I didnt look carefully enough to realize that. If you dont mind me asking one more dumb question: Who/what decides whether the 2 speaker amps go on the same link or different links? Is it purely based on board design?

Reply:
If you put two speakers on the same link, you increase the capacitive load and make the routing more complicated (all devices need to see the same clocks, otherwise you could have electrical issues).
If you put two speakers on different links, you can potentially simplify the routing with different wires going in different directions, BUT you will now need to rely on a synchronization of the two links with non-standard means (we have a 'gsync' hw signal on Intel platforms but that's not part of the MIPI standard). You also need to use 2 additional pins per link.
Also with more than one link, you can potentially make power management simpler, e.g. you only enable the links that are active, and go in clock-stop mode if not.
In short, the decision is going to be a compromise between hardware/software/routing/pin count considerations. I think for 2 speakers it's likely the same complexity overall. If you start adding more devices then at some point electrical limitations and bandwidth restrictions will force the use of more than one link.
Does this help?

Reply:
Yes, thank you very much!